cocos/platform/CCFileUtils.cpp

```diff
@@ -1226,6 +1226,14 @@ std::vector<std::string> FileUtils::listFiles(const std::string& dirPath) const
     return files;
 }
 
+void FileUtils::listFiles(const std::string& dirPath, std::function<void(std::vector<std::string>)> callback) const
```

Comment:
I would like to keep the callback by value, so that users can better leverage c++ 11 move semantics to avoid copies. 

If we take the callback by const reference (const &), then when we assign it into our execution queue to be executed off-thread with the AsyncTask pool, we will be making a copy (since it is const and we cannot mutate it). 

If we take the callback by l-value reference (&), then we can call std::move on it to avoid a copy. However, then our code will not accept r-value callbacks. Example:

```
Async::enque(std::function<void(void)>& callback)
{

}

void foo() 
{
auto bar = [] () { };
Async::enque(bar); // Compiles, will not copy bar.
Async::enque([] () {}); // Fails to compile, lambda is an r-value.
}
```
Likewise, taking the callback by r-value reference (&&), we will have the opposite problem (we cannot accept bar without std::move'ing into it, which we may not always want to do.)

**However**, we can have the best of both worlds by taking by value, and using std::move! Example:

```
Async::queue(std::function<void(void)> callback) 
{
    // If I took a const reference, I couldn't std::move here, and would have to copy.
    _queue.push_back(std::move(callback)); // This is safe because I own callback.
}

foo() {
auto lambda1 = [] () { };
auto lambda2 = [] () { };
Async::enque(lambda1) // This will copy, BUT I save a copy by calling std::move in enque. 
Net result is I have the same number of copies as if I took a const reference. 
Async::enque(std::move(lambda2)); // This will move, much cheaper then copy. 
Async::enque([] () {}); // r-value, avoids copy.
}
```
As you can see, even if the user forgets to call std::move at the call site, I have the same number of copies that I would if I took the callback by const reference. And if the user calls std::move at the function callsite, we can make no copies at all! We get the benefits of using c++ objects on the stack, with traceable lift times, but we have the performance of code that is only performing pointer copies, and not expensive copy constructors. I feel that cocos2d-x could better leverage move semantics as a c++11 game engine, so it's good to have examples on how to use move semantics in the code base.